pkgs/build-support/cc-wrapper/ld-wrapper.sh

```diff
@@ -84,49 +85,60 @@ if [ "$NIX_DONT_SET_RPATH" != 1 ]; then
         fi
     }
 
+    declare -A libs
+
+    addToLibs() {
+        libs["lib$1.so"]=1
+    }
+
     declare -A rpathsSeen
-    declare -a rpaths
 
-    addToRPath() {
+    addRPath() {
         # If the path is not in the store, don't add it to the rpath.
         # This typically happens for libraries in /tmp that are later
         # copied to $out/lib.  If not, we're screwed.
         if [ "${1:0:${#NIX_STORE}}" != "$NIX_STORE" ]; then return 0; fi
         if [[ -z ${rpathsSeen[$1]} ]]; then
-            rpaths+=("$1")
+            extra+=(-rpath "$1")
             rpathsSeen[$1]=1
         fi
     }
 
-    declare -a libs
-
     # First, find all -L... switches.
     allParams=("${params[@]}" ${extra[@]})
+    nParams=${#allParams[@]}
     n=0
-    while [ $n -lt ${#allParams[*]} ]; do
+    while [ $n -lt $nParams ]; do
         p=${allParams[n]}
         p2=${allParams[$((n+1))]}
-        if [ "${p:0:3}" = -L/ ]; then
-            addToLibPath ${p:2}
-        elif [ "$p" = -L ]; then
-            addToLibPath ${p2}
-            n=$((n + 1))
-        elif [ "$p" = -l ]; then
-            libs+=(${p2})
-            n=$((n + 1))
-        elif [ "${p:0:2}" = -l ]; then
-            libs+=(${p:2})
-        elif [ "$p" = -dynamic-linker ]; then
-            # Ignore the dynamic linker argument, or it
-            # will get into the next 'elif'. We don't want
-            # the dynamic linker path rpath to go always first.
-            n=$((n + 1))
-        elif [[ "$p" =~ ^[^-].*\.so($|\.) ]]; then
-            # This is a direct reference to a shared library, so add
-            # its directory to the rpath.
-            path="$(dirname "$p")";
-            addToRPath "${path}"
-        fi
+        case "$p" in
+            -L)
+                addToLibPath "${p2}"
+                n=$((n + 1))
+                ;;
+            -L/*)
+                addToLibPath "${p:2}"
+                ;;
+            -l)
+                addToLibs "${p2}"
+                n=$((n + 1))
+                ;;
+            -l*)
+                addToLibs "${p:2}"
+                ;;
+            -dynamic-linker)
+                # Ignore the dynamic linker argument, or it
+                # will get into the next clause. We don't want
+                # to add the dynamic linker path to rpath.
+                n=$((n + 1))
+                ;;
+            [^-]*.so | [^-]*.so.*)
```

Comment:
Would `?` in `[^-]*.so | [^-]*.so.?` replicate the original regex more accurately?